# main.py
import sys
from PyQt5.QtWidgets import (QApplication, QMainWindow, QMessageBox,
                             QTableWidgetItem, QHeaderView)
from ui.ui_placement import Ui_MainWindow
from data.parser import parse_table_sites, parse_table_zones
from model.optimizer import optimize_placement


class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)

        #configuration des tableaux
        self.setup_tables()

        #connexion des signaux
        self.connect_signals()

        #charger des donnÃ©es exemples
        self.load_example_data()

        #message de bienvenue dans la barre de statut
        self.statusBar().showMessage("PrÃªt Ã  optimiser le placement des ambulances")

    def setup_tables(self):
        #table Sites
        self.ui.tableSites.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
        self.ui.tableSites.setAlternatingRowColors(True)
        self.ui.tableSites.verticalHeader().setVisible(True)

        #table Zones
        self.ui.tableZones.horizontalHeader().setSectionResizeMode(QHeaderView.Stretch)
        self.ui.tableZones.setAlternatingRowColors(True)
        self.ui.tableZones.verticalHeader().setVisible(True)

        #zone de resultats en lecture seule
        self.ui.textResults.setReadOnly(True)

    def connect_signals(self):
        """Connecte tous les signaux aux slots"""
        #bouton d'optimisation
        self.ui.btnOptimize.clicked.connect(self.run_optimizer)

        #boutons pour les sites
        self.ui.btnAddSite.clicked.connect(self.add_site_row)
        self.ui.btnRemoveSite.clicked.connect(self.remove_site_row)

        #boutons pour les zones
        self.ui.btnAddZone.clicked.connect(self.add_zone_row)
        self.ui.btnRemoveZone.clicked.connect(self.remove_zone_row)

    def add_site_row(self):
        """Ajoute une nouvelle ligne dans le tableau des sites"""
        row_count = self.ui.tableSites.rowCount()
        self.ui.tableSites.insertRow(row_count)

        #initialiser les cellules avec des valeurs par dÃ©faut
        self.ui.tableSites.setItem(row_count, 0, QTableWidgetItem(f"Site_{row_count + 1}"))
        self.ui.tableSites.setItem(row_count, 1, QTableWidgetItem("0"))
        self.ui.tableSites.setItem(row_count, 2, QTableWidgetItem("0"))
        self.ui.tableSites.setItem(row_count, 3, QTableWidgetItem(""))

        self.statusBar().showMessage(f"Site ajoutÃ© (ligne {row_count + 1})", 3000)

    def remove_site_row(self):
        """Supprime la ligne sÃ©lectionnÃ©e dans le tableau des sites"""
        current_row = self.ui.tableSites.currentRow()
        if current_row >= 0:
            self.ui.tableSites.removeRow(current_row)
            self.statusBar().showMessage(f"Site supprimÃ©", 3000)
        else:
            QMessageBox.warning(self, "Attention", "Veuillez sÃ©lectionner une ligne Ã  supprimer")

    def add_zone_row(self):
        """Ajoute une nouvelle ligne dans le tableau des zones"""
        row_count = self.ui.tableZones.rowCount()
        self.ui.tableZones.insertRow(row_count)

        #initialiser les cellules avec des valeurs par dÃ©faut
        self.ui.tableZones.setItem(row_count, 0, QTableWidgetItem(f"Zone_{row_count + 1}"))
        self.ui.tableZones.setItem(row_count, 1, QTableWidgetItem("0"))
        self.ui.tableZones.setItem(row_count, 2, QTableWidgetItem("0"))
        self.ui.tableZones.setItem(row_count, 3, QTableWidgetItem("1000"))
        self.ui.tableZones.setItem(row_count, 4, QTableWidgetItem("1"))

        self.statusBar().showMessage(f" Zone ajoutÃ©e (ligne {row_count + 1})", 3000)

    def remove_zone_row(self):
        """Supprime la ligne sÃ©lectionnÃ©e dans le tableau des zones"""
        current_row = self.ui.tableZones.currentRow()
        if current_row >= 0:
            self.ui.tableZones.removeRow(current_row)
            self.statusBar().showMessage(f"Zone supprimÃ©e", 3000)
        else:
            QMessageBox.warning(self, "Attention", "Veuillez sÃ©lectionner une ligne Ã  supprimer")

    def load_example_data(self):
        """Charge des donnÃ©es d'exemple pour faciliter les tests"""
        # Exemple 1 : - HÃ´pital Centre couvre TOUTES les zones (5km max) et Base Est couvre B et E (zones peuplÃ©es)
        sites_data = [
            ("ðŸ¥ HÃ´pital Centre", 5.0, 5.0, "3"),
            ("ðŸš’ Caserne Nord", 2.0, 8.0, "2"),
            ("ðŸ¢ Station Sud", 8.0, 2.0, "2"),
            ("ðŸ­ Base Est", 10.0, 5.0, "3"),
        ]
        zones_data = [
            ("ðŸ˜ï¸ Quartier A", 3.0, 6.0, "5000", "1"),
            ("ðŸ˜ï¸ Quartier B", 7.0, 7.0, "8000", "2"),
            ("ðŸ˜ï¸ Quartier C", 6.0, 3.0, "6000", "1"),
            ("ðŸ˜ï¸ Quartier D", 1.0, 2.0, "4000", "3"),
            ("ðŸ˜ï¸ Quartier E", 9.0, 8.0, "7000", "1"),
        ]
        #2 Exemple 2 : CAS "ZONE ISOLÃ‰E" : Quartier D loin de tout
        """sites_data = [
            ("ðŸ¥ HÃ´pital Centre", 5.0, 5.0, "3"),
            ("ðŸš’ Caserne Nord", 2.0, 8.0, "2"),
            ("ðŸ¢ Station Sud", 8.0, 2.0, "2"),
            ("ðŸ­ Base Est", 10.0, 5.0, "3"),
        ]
        # MÃªmes sites
        zones_data = [
            ("ðŸ˜ï¸ Quartier A", 3.0, 6.0, "5000", "1"),
            ("ðŸ˜ï¸ Quartier B", 7.0, 7.0, "8000", "2"),
            ("ðŸ˜ï¸ Quartier C", 6.0, 3.0, "6000", "1"),
            ("ðŸ˜ï¸ Quartier D", 15.0, 15.0, "4000", "3"),  # TRÃˆS LOIN !
            ("ðŸ˜ï¸ Quartier E", 9.0, 8.0, "7000", "1"),
        ]"""
        # Exemple 3 : D devient le plus prioritaire !! donc A et D auront la mm imprtance
        """sites_data = [
            ("ðŸ¥ HÃ´pital Centre", 5.0, 5.0, "3"),
            ("ðŸš’ Caserne Nord", 2.0, 8.0, "2"),
            ("ðŸ¢ Station Sud", 8.0, 2.0, "2"),
            ("ðŸ­ Base Est", 10.0, 5.0, "3"),
        ]
        # MÃªmes sites
        zones_data = [
            ("ðŸ˜ï¸ Quartier A", 3.0, 6.0, "5000", "1"),
            ("ðŸ˜ï¸ Quartier B", 7.0, 7.0, "8000", "2"),
            ("ðŸ˜ï¸ Quartier C", 6.0, 3.0, "6000", "1"),
            ("ðŸ˜ï¸ Quartier D", 1, 2, "5000", "5"),  # TRÃˆS PRIORITAIRE!
            ("ðŸ˜ï¸ Quartier E", 9.0, 8.0, "7000", "1"),
        ]"""
        #exemple 4: mm que exemple 1 mais rayon 3



        self.ui.tableSites.setRowCount(len(sites_data))
        for i, (name, x, y, cap) in enumerate(sites_data):
            self.ui.tableSites.setItem(i, 0, QTableWidgetItem(name))
            self.ui.tableSites.setItem(i, 1, QTableWidgetItem(str(x)))
            self.ui.tableSites.setItem(i, 2, QTableWidgetItem(str(y)))
            self.ui.tableSites.setItem(i, 3, QTableWidgetItem(cap))

        # Exemple de zones


        self.ui.tableZones.setRowCount(len(zones_data))
        for i, (name, x, y, pop, prio) in enumerate(zones_data):
            self.ui.tableZones.setItem(i, 0, QTableWidgetItem(name))
            self.ui.tableZones.setItem(i, 1, QTableWidgetItem(str(x)))
            self.ui.tableZones.setItem(i, 2, QTableWidgetItem(str(y)))
            self.ui.tableZones.setItem(i, 3, QTableWidgetItem(pop))
            self.ui.tableZones.setItem(i, 4, QTableWidgetItem(prio))

        # ParamÃ¨tres par dÃ©faut
        self.ui.spinTotalAmbulances.setValue(6)
        self.ui.spinMaxDistance.setValue(5.0)

    def run_optimizer(self):
        """Lance l'optimisation et affiche les rÃ©sultats"""
        try:
            #rÃ©cupÃ©ration des donnÃ©es
            sites = parse_table_sites(self.ui.tableSites)
            zones = parse_table_zones(self.ui.tableZones)
            total_ambulances = self.ui.spinTotalAmbulances.value()
            max_distance = self.ui.spinMaxDistance.value()

            #dÃ©terminer le mode
            mode_text = self.ui.comboModeDecision.currentText()
            if 'Binaire' in mode_text or 'binaire' in mode_text:
                mode = 'binaire'
            else:
                mode = 'entier'

            # Validation
            if not sites:
                QMessageBox.warning(self, "âš Erreur", "Veuillez ajouter au moins un site")
                return

            if not zones:
                QMessageBox.warning(self, "âš Erreur", "Veuillez ajouter au moins une zone")
                return

            #message de progression
            self.ui.textResults.setText("Optimisation en cours...\n\nCalcul du placement optimal...")
            self.statusBar().showMessage("Optimisation en cours...")
            QApplication.processEvents()  # Forcer la mise Ã  jour de l'interface

            #lancer l'optimisation
            placement, stats = optimize_placement(sites, zones, total_ambulances,
                                                  max_distance, mode)

            #vÃ©rifier les erreurs
            if 'error' in stats:
                QMessageBox.critical(self, " Erreur d'optimisation", stats['error'])
                self.ui.textResults.setText(f" ERREUR\n\n{stats['error']}")
                self.statusBar().showMessage("Erreur d'optimisation", 5000)
                return

            #afficher les rÃ©sultats
            self.display_results(placement, stats, mode, max_distance, total_ambulances)
            self.statusBar().showMessage("Optimisation terminÃ©e avec succÃ¨s!", 5000)

        except Exception as e:
            error_msg = f"Une erreur s'est produite:\n{str(e)}"
            QMessageBox.critical(self, "Erreur", error_msg)
            self.ui.textResults.setText(f"ERREUR\n\n{str(e)}")
            self.statusBar().showMessage("Erreur", 5000)

    def display_results(self, placement, stats, mode, max_distance, total_ambulances):
        """Affichage moderne, clair et professionnel avec les vraies probabilitÃ©s"""

        lines = []
        add = lines.append

        # === EN-TÃŠTE ===
        add("RÃ‰SULTATS DE L'OPTIMISATION PROBABILISTE")
        add("â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€")
        add(f"  Mode               : {mode.capitalize()}")
        add(f"  Distance maximale  : {max_distance:.1f} km")
        add(f"  Budget ambulances  : {total_ambulances}")
        add("")

        # === PLACEMENT ===
        add("PLACEMENT OPTIMAL DES AMBULANCES")
        add("â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€")
        if placement:
            for site, nb in sorted(placement.items(), key=lambda x: -x[1]):
                if nb > 0:
                    add(f"  {site:<35} â†’ {nb} ambulance{'s' if nb > 1 else ' '}")
            add("")
            add(f"  Total utilisÃ© : {stats['total_ambulances_placed']} / {total_ambulances} ambulances")
        else:
            add("  Aucune ambulance placÃ©e")
        add("")

        # === COUVERTURE PROBABILISTE (le vrai truc impressionnant) ===
        add("COUVERTURE RÃ‰ELLE ATTENDUE (ambulances parfois occupÃ©es)")
        add("â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€")

        # Taux global
        add(f"  Couverture moyenne attendue : {stats['coverage_percentage']}%")
        add("")

        # DÃ©tail par zone (les 10 plus importantes + rÃ©sumÃ©)
        add("DÃ©tail par zone (ambulances couvrantes â†’ probabilitÃ© de sauvetage)")
        add("â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€")

        # Tri par contribution dÃ©croissante
        sorted_zones = sorted(stats['details_zones'], key=lambda x: x['contribution'], reverse=True)

        for zone in sorted_zones[:12]:  # les 12 plus importantes
            nom = zone['nom']
            pop = f"{zone['population']:,}".replace(",", " ")
            amb = zone['ambulances']
            proba = zone['probabilite_%']
            add(f"  {nom:<30} ({pop} hab.) â†’ {amb} amb. â†’ {proba}% de chance dâ€™Ãªtre sauvÃ©")

        if len(sorted_zones) > 12:
            add(f"  ... et {len(sorted_zones) - 12} autres zones")
        add("")

        # === RÃ‰SUMÃ‰ FINAL ===
        add("RÃ‰SUMÃ‰")
        add("â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€")
        add(f"  Population sauvÃ©e en moyenne : {stats['population_covered']:,} habitants")
        add(f"  ProbabilitÃ© moyenne de sauvetage : {stats['coverage_percentage']:.1f} %")
        add("")
        add("ModÃ¨le basÃ© sur une disponibilitÃ© rÃ©elle de 60 % (ambulances parfois en intervention)")

        # Affichage final
        result_text = "\n".join(lines)
        self.ui.textResults.setPlainText(result_text)
        self.ui.textResults.setFont(self.ui.textResults.font())  # police propre


def main():
    """point d'entrÃ©e de l'application"""
    app = QApplication(sys.argv)

    #style de l'application
    app.setStyle('Fusion')

    #crÃ©er et afficher la fenÃªtre
    window = MainWindow()
    window.show()

    sys.exit(app.exec_())


if __name__ == "__main__":
    main()